tudatBundle/pagmo2/include/pagmo/algorithms/giACOmo.hpp

```diff
@@ -0,0 +1,1017 @@
+ *
+ * See:  M. Schlueter, et al. (2009). Extended ant colony optimization for non-convex
+ * mixed integer non-linear programming. Computers & Operations Research.
+ */
+class gi_aco_mo
+{
+public:
+    /// Single entry of the log (gen, fevals, ideal_point)
+    typedef std::tuple<unsigned, unsigned long long, vector_double> log_line_type;
+    /// The log
+    typedef std::vector<log_line_type> log_type;
+
+    /// Constructors
+    /**
+    * Empty constructor
+    * */
+
+    // gi_aco_mo( ){ }
+
+    /**
+    * Constructs the ACO user defined algorithm for multi-objective optimization.
+    *
+    * @param[in] gen Generations: number of generations to evolve.
+    * @param[in] acc Accuracy parameter: for inequality and equality constraints .
+    * @param[in] fstop Objective stopping criterion: when the objective value reaches this value, the algorithm is stopped [for multi-objective, this applies to the first obj. only].
+    * @param[in] impstop Improvement stopping criterion: if a positive integer is assigned here, the algorithm will count the runs without improvements, if this number will exceed IMPSTOP value, the algorithm will be stopped.
+    * @param[in] evalstop Evaluation stopping criterion: same as previous one, but with function evaluations
+    * @param[in] focus Focus parameter: this parameter makes the search for the optimum greedier and more focused on local improvements (the higher the greedier). If the value is very high, the search is more focused around the current best solutions
+    * @param[in] ker Kernel: number of solutions stored in the solution archive
+    * @param[in] oracle Oracle parameter: this is the oracle parameter used in the penalty method
+    * @param[in] paretomax Max number of non-dominated solutions: this regulates the max number of Pareto points to be stored
+    * @param[in] epsilon Pareto precision: the smaller this parameter, the higher the chances to introduce a new solution in the Pareto front
+    * @param seed seed used by the internal random number generator (default is random)
+    * @throws std::invalid_argument if \p acc is not \f$ \in [0,1[\f$, \p impstop is not a
+    * positive integer, \p evalstop is not a positive integer, \p focus is not \f$ \in [0,1[\f$, \p ants is not a positive integer,
+    * \p ker is not a positive integer, \p oracle is not positive, \p paretomax is not a positive integer, \p epsilon is not \f$ \in [0,1[\f$
+    */
+    gi_aco_mo(unsigned gen = 100u, double acc = 0.95, double  fstop = 0.0000001, int impstop = 100, int evalstop = 100,
+          double focus = 10,  unsigned ker = 500, double oracle=1000, unsigned paretomax = 10,
+            double epsilon = 0.9, unsigned seed = pagmo::random_device::next())
+        : m_gen(gen), m_acc(acc), m_fstop(fstop), m_impstop(impstop), m_evalstop(evalstop), m_focus(focus),
+          m_ker(ker), m_oracle(oracle), m_paretomax(paretomax), m_epsilon(epsilon), m_e(seed), m_seed(seed), m_verbosity(0u),
+          m_log()
```

Comment:
Giacomo

1) Do not use islands to evolve. Use directly the algo evolve method. 
2) Do not use print to file to debug. Use the screen streams.
3) Start using rosenbrock 10 as a problem.

Reply:
```
// Algorithm (setting iterations to 100)
pagmo::algorithm algo{ gi_aco_mo{100} };

// Problem
pagmo::problem prob{rosenbrock{1000}};

// Population
pagmo::population pop{algo, prob, 10};

// Evolve for 1000 generations
pop = algo.evolve(pop)

// Print current optimum to console
std::cout << pop << std::endl;

}
```

Reply:
The above code is untested

Reply:
Sure for the name. g_aco with acronym G-ACO looks fine to me.

1) I will change the evolve, but can I ask you why  I should not use the island?

2) I was printing to a file because I was using MATLAB to plot the data and see whether there was improvement. For screen streams, do you mean to look at the application output in my C++ platform?

Reply:
1 - The islands open up threads and forks and you may end up debugging your code for thread safety rather than its content. Also error messages may be obscured when the algo evolve is called via island. Since you are developing its better to focus on calling as directly as possible the new methods developed
2 - The best chain is to code a log, to then have on screen something like:

```
  Gen:        Fevals:          Best:             F:            CR:       Variant:            dx:            df:
      1             20        179.464       0.675185       0.326315             15        9.64601         815242
      2             40        179.464       0.675185       0.326315             15        9.03224         111333
      3             60         10.793       0.520159       0.799299             17         8.2787        78644.1
      4             80        8.73417       0.447463       0.559056             15        7.09521        44740.9
      5            100        8.73417       0.447463       0.559056             15        3.68142        6981.56
      6            120        8.73417       0.447463       0.559056             15        3.68142        6981.56
      7            140        8.73417       0.447463       0.559056             15        8.02873        2488.06
      8            160        2.62825       0.929304       0.184143              7        10.0754        2494.17
      9            180        2.62825       0.929304       0.184143              7        7.52991        1403.77
     10            200        2.62825       0.929304       0.184143              7        7.52991        1403.77
     11            220        2.62825       0.929304       0.184143              7        7.52991        1403.77
     12            240        2.62825       0.929304       0.184143              7        7.52991        1403.77
     13            260        2.62825       0.929304       0.184143              7        8.11855        675.195
     14            280        1.29047       0.926908       0.193483              1        8.22928        676.533
     15            300        1.29047       0.926908       0.193483              1        3.18796        258.864
     16            320        1.29047       0.926908       0.193483              1        3.18796        258.864
     17            340       0.118338      -0.832508       -0.35505             16        3.36788        260.036
     18            360       0.118338      -0.832508       -0.35505             16        2.91671        136.045
     19            380       0.118338      -0.832508       -0.35505             16        2.91671        136.045
     20            400       0.118338      -0.832508       -0.35505             16        2.91671        136.045
     21            420       0.118338      -0.832508       -0.35505             16        2.91671        136.045
     22            440       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     23            460       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     24            480       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     25            500       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     26            520       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     27            540       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     28            560       0.118338      -0.832508       -0.35505             16        1.04414        116.153
     29            580       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     30            600       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     31            620       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     32            640       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     33            660       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     34            680       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     35            700       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     36            720       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     37            740       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     38            760       0.118338      -0.832508       -0.35505             16        8.64696        103.929
     39            780       0.118338      -0.832508       -0.35505             16        10.7329         59.328
     40            800       0.118338      -0.832508       -0.35505             16        10.7329         59.328
     41            820       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     42            840       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     43            860       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     44            880       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     45            900       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     46            920       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     47            940       0.118338      -0.832508       -0.35505             16         8.5833        50.3479
     48            960       0.118338      -0.832508       -0.35505             16        9.85526        25.0998
     49            980       0.118338      -0.832508       -0.35505             16        9.85526        25.0998
     50           1000       0.118338      -0.832508       -0.35505             16        9.85526        25.0998
```

Do not use MATLAB. The more layers you add the more you will suffer. Just debug C++ code using c++ screen output. The example above is for rosenbrock 10 and de1220

Reply:
LAter we will be able to test and plot in python

Reply:
The constructor of population does not seem to match the piece of code:
// Population
pagmo::population pop{algo, prob, 10};

but pop seems to want as input the problem, the population and the seed. However, how can I then link the problem to the algorithm?

Reply:
Sorry, in the above comment I meant: the problem, the population SIZE and the seed

Reply:
Sceki, as detailed in the online documentation the population can be constructed from a problem:
```pagmo::population pop{prob, 10};```

then, as written above:

```
pop  =algo.evolve(pop)
```
The link is there as pop is passed to algo.

Reply:
Thank you very much!